src/FakeItEasy/Core/CallOrderingContext.cs

```diff
@@ -0,0 +1,87 @@
+﻿namespace FakeItEasy.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    internal class CallOrderingContext : ICallOrderingContext
+    {
+        private readonly CallWriter callWriter;
+        private readonly HashSet<FakeManager> fakeManagers;
+        private readonly List<AssertedCall> assertedCalls;
+        private int currentSequenceNumber;
+
+        public CallOrderingContext(CallWriter callWriter)
+        {
+            this.callWriter = callWriter;
+            this.currentSequenceNumber = -1;
+            this.fakeManagers = new HashSet<FakeManager>();
+            this.assertedCalls = new List<AssertedCall>();
+        }
+
+        public void CheckNextCall(FakeManager fakeManager, Func<IFakeObjectCall, bool> callPredicate, string callDescription, Repeated repeatConstraint)
+        {
+            this.fakeManagers.Add(fakeManager);
+            this.assertedCalls.Add(new AssertedCall { CallDescription = callDescription, RepeatDescription = repeatConstraint.ToString() });
+            var allCalls = this.fakeManagers.SelectMany(f => f.AllRecordedCalls.Cast<IFakeObjectCall>()).OrderBy(c => c.SequenceNumber).ToList();
+            var calls = allCalls
+                .Where(callPredicate)
+                .SkipWhile(c => c.SequenceNumber <= this.currentSequenceNumber)
+                .ToList();

```

Comment:
Alternatively, an move the `SkipWhile` and the `Where` predicates inside the `SelectMany`, to make the `OrderBy` cheaper because it doesn't have to sort all the items that we will eventually discard…